packages/yarnpkg-core/tests/Configuration.test.ts

```diff
@@ -332,9 +359,304 @@ describe(`Configuration`, () => {
+      const mapShape = map.get(`foo`) as Map<string, any>;
+      expect(mapShape.get(`number`)).toBe(0);
+
+      const array = configuration.get(`array`) as Array<string>;
+      expect(array[0]).toBe(`1`);
+      expect(array[1]).toBe(`2`);
+    });
+  });
+
+  it(`it should correctly extend or reset the values`, async() => {
+    const {plugins, pluginConfiguration} = await initConfigurationPlugin(`{
+      string: {
+        description: "",
+        type: "STRING",
+        default: "",
+      },
+      string2: {
+        description: "",
+        type: "STRING",
+        default: "",
+      },
+      stringArray: {
+        description: "",
+        type: "STRING",
+        isArray: true,
+        default: [],
+      },
+      stringArray2: {
+        description: "",
+        type: "STRING",
+        isArray: true,
+        default: [],
+      },
+      shape: {
+        description: "",
+        type: "SHAPE",
+        properties: {
+          number: {
+            description: "",
+            type: "NUMBER",
+            default: 0,
+          },
+          string: {
+            description: "",
+            type: "STRING",
+            default: "default",
+          },
+        },
+      },
+      shape2: {
+        description: "",
+        type: "SHAPE",
+        properties: {
+          number: {
+            description: "",
+            type: "NUMBER",
+            default: 0,
+          },
+          string: {
+            description: "",
+            type: "STRING",
+            default: "default",
+          },
+        },
+      },
+      map: {
+        description: "",
+        type: "MAP",
+        valueDefinition: {
+          description: "",
+          type: "SHAPE",
+          properties: {
+            number: {
+              description: "",
+              type: "NUMBER",
+              default: 0,
+            },
+            string: {
+              description: "",
+              type: "STRING",
+              default: "default",
+            },
+          },
+        },
+      },
+      map2: {
+        description: "",
+        type: "MAP",
+        valueDefinition: {
+          description: "",
+          type: "SHAPE",
+          properties: {
+            number: {
+              description: "",
+              type: "NUMBER",
+              default: 0,
+            },
+            string: {
+              description: "",
+              type: "STRING",
+              default: "default",
+            },
+          },
+        },
+      },
+    }`);
+    await initializeConfiguration({
+      plugins,
+      string: `bar`,
+      string2: {value: `bar2`, strategy: `reset`},
+      stringArray: [`bar`],
+      stringArray2: {value: [`bar2`], strategy: `reset`},
+      shape: {
+        number: 2,
+        string: `bar`,
+      },
+      shape2: {
+        value: {
+          number: 22,
+          string: `bar2`,
+        },
+        strategy: `reset`,
```

Comment:
nit: not a fan of `strategy: reset`, the "strategy" here doesn't look super clear to me ... `onConflict: reset`, maybe? 🤔

And do we need the `value` field? Perhaps we could just use the remaining keys as the proper object; it would allow to remove an indentation level if someone wanted to use `onConflict: reset` at the top-level to disable the yarnrc inheritance altogether:

```yaml
onConflict: reset

npmRegistryServer:
  # ...
```

Reply:
> And do we need the value field? Perhaps we could just use the remaining keys as the proper object

I think this will make configuration easier, but I'm not sure how this works in `ARRAY` and `SINGLEVALUE`, would you be willing to tell me more 🙏?

```
// MAP & SHAPE
onConflict: "xxx"
packageExtensions:
  onConflict: "xxx"
  gatsby-cli@*:
    onConflict: "xxx"
    peerDependencies:
      onConflict: "xxx"
      eslint: "*"
      gatsby: "*"
      
// ARRAY
🤔

// SINGLE VALUE (string, number, boolean)
🤔
/* 
(Currently, if a single value has the `reset' option, it will still overwrite the original value, even if `useWithSource.opts.overwrite === false`, but this may be an overdesign 🧐.)
*/
```

Reply:
Yes, in the case of array we'd probably need to have a `value` field, I don't see a clear way to avoid it there... 🤔

It's a first iteration anyway, we can improve that later on if we find a prettier syntax

> (Currently, if a single value has the `reset` option, it will still overwrite the original value, even if `useWithSource.opts.overwrite === false`, but this may be an overdesign 🧐.)

Looking at its usage, we probably should remove `useWithSource.opts.overwrite` entirely, and replace it by adding `onConflict: reset` to all the places that used it.

Reply:
These days I have done some things

1. I removed `useWithSource.opts.overwrite` and `BaseSettingsDefinition.concatenateValues` (It doesn't seem to matter much, and it doesn't seem to work without `overwrite` 🧐)

2. I implemented `onConflict` and it now has three behaviors

    #### extend(default)
    attempt to merge or overwrite (reset)

    #### reset
    overwrite a field whether it has a value or not

    #### skip
    if the value already exists in the field, it will not be parsed

3. I modified the test to cover the onConflict behavior